### ISSUE TITLE
componentWillUnmount is never called for modals

### ISSUE THREAD

**Issue by madvas** [wix/react-native-navigation#39]
I simply added those lines to https://github.com/wix/react-native-navigation/blob/master/example/src/screens/ModalScreen.js 

```
componentWillMount () {
    console.log("modal mounted");
  }
componentWillUnmount () {
  console.log("modal unmounted");
}
```

`componentWillMount` is called everytime modal is shown. But `componentWillUnmount` isn't called when modal is dismissed.

**Comment by alexeypodolian** [comment]
Faced with this issue too.

**Comment by webzepter** [comment]
Also bumped into this issue.

**Comment by maxs15** [comment]
+1 Same for the pushed screens

**Comment by alexeypodolian** [comment]
@guyca Any updates with this issue?

**Comment by alexeypodolian** [comment]
@guyca 
Do we really need to call `finish()` in `RootActivity` in `onPause()` method? 

```
    @Override
    protected void onPause() {
        super.onPause();
        finish();
    }
```

**Comment by guyca** [comment]
@alexeypodolian Screens don't get unmounted when another screen gets pushed or a modal is displayed by design. We want to keep them intact so when the pushed screen is popped the transition will be seamless and the screen's previous "navigational" state is persisted.

For example; Screen A contains a large list of products, clicking an item pushes Screen B with product details. User scrolls down the list and a few items are loaded into the list. The user then clicks an item and screen B is pushed. Once we pop screen B - our scroll position in screen A is maintained.
When screen B is displayed - screen A doesn't get unmounted. When screen B is removed (we popped it) it DOES get unmounted since we can't go back it it.

We understand the problem this behaviour created and will add Life Cycle callbacks to all screens.

**Comment by webzepter** [comment]
@guyca Did you correctly understand the issue? We meant unmounting component on **DISMISSING** a modal, not pushing a new.

**Comment by guyca** [comment]
@webzepter I'll check it out. Hopefully be the end of next week.

**Comment by epaga** [comment]
@guyca getting this issue as well - any way around this? Seems like this might be more a thing for react-native-controllers, right? - RCCManager.unregisterController doesn't make sure the underlying react component is unmounted (not experienced enough with the RCC API to know how to do this or I would provide a pull request)...

**Comment by guyca** [comment]
Seems to work correctly on Android. @gran33 can you take this?

**Comment by DanielZlotin** [comment]
Fixed in `react-native-controllers@2.0.9`.

Doing `npm install` should remedy the issue.
The bug was the view inside RCTViewController was not being dealloc'd; this happened only when we switched the app root when we have modals showing.
